### ISSUE TITLE
[question][2.0] exports_sources with .gitignore support

### ISSUE THREAD

**Issue by Da-LiFe** [conan-io/conan#12238]
I'm currently starting to port conan 1.x recipes in our organization to conan 2.0 syntax using conan 1.52.0. One pattern we use in most of our projects is this syntax for exports_sources in order not to export files ignored by git:

`exports_sources = ["*", "!.gitignore"] + ["!%s" % x for x in tools.Git().excluded_files()]`

But apparently, support for excluded_files() is not provided in [conan.tools.scm.git](https://github.com/conan-io/conan/blob/1.52.0/conan/tools/scm/git.py). Besides that, #4023 proposed to use the SCM feature, which is also [deprecated](https://docs.conan.io/en/1.52/migrating_to_2.0/recipes.html#the-scm-attribute). 

As of now, I don't see a chance to migrate this behaviour to conan 2.0 except reimplementing it ourselfs. Is there another solution I've overseen? Is there a chance to make `excluded_files()` available in `conan.tools.scm.git` or is there a rationale behind removing it? Or asking more general: Will there be a mechanism in exports_sources to not copy files ignored by git?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
It might be possible to implement the ``excluded_files`` in the new ``Git`` helper, but probably it is not even necessary?

The recommended way for anything that is relatively complex, like your expression, is to use the ``def export_sources(self)`` method instead of the attribute. There, it could be something as:

```python
def export_sources(self):
     git = Git(self, self.recipe_folder)
     scm_url, scm_commit = git.get_url_and_commit()
     git = Git(self, self.export_sources_folder)
     git.clone(url=scm_url, target=".")
     git.checkout(commit=scm_commit)
```

I am probably lacking some git-fu, but if there is a git command that allows a clean copy of the files into another folder, maybe it is just one ``git.run("archive ....")``. 

To be clear, I am not opposed to having the "exclude_files" feature, but it seems this is hiding a use case that is: "I want to export the sources of my repo that aren't gitignored". Maybe running a full ``copy(self, "*", ....)`` and then running ``git.run("clean -xdf")`` to remove files? I am looking for something that is easier to read and understand than ``["*", "!.gitignore"] + ["!%s" % x for x in tools.Git().excluded_files()]``

**Comment by Da-LiFe** [comment]
The use case is: "Copy all tracked and untracked files of the current repo except the ignored ones". This should avoid copying e.g. large files that aren't used for packaging anyways. Git clone will neither copy changed nor untracked files, so this is not an option for our use case I think. Copy first and clean afterwards won't give any increase in speed. It's more like a filesystem copy with filter I would say and that's probably nothing one can do easily with just git commands.

But I agree, using a method can increase readability and is anyways required, because the Git constructor now requires `self`:

```python
    @property
    def exports_sources(self):
        from conan.tools.scm import Git

        # copy all files in the repo except .gitignore
        file_list = ["*", "!.gitignore"]
        # but filter all files marked as ignored by git 
        file_list += [f"!{file}" for file in Git(self).excluded_files()]
        # let conan copy the files to the cache
        return file_list
```

Using `@property` I just do not have to deal with `copy(self, ....)`. 

For our use case I cannot find something more elegant right now and having `Git(self).excluded_files()` in the API would be quite helpful.

**Comment by memsharded** [comment]
I am proposing in https://github.com/conan-io/conan/pull/12246 a different approach, in the user side:

```python
 def export_sources(self):
       git = Git(self)
       for i in git.included_files():
             shutil.copy2(i, self.export_sources_folder)
```

On the Conan side, implementation is straightforward too, which is also very important for maintenance:
```python
    def included_files(self):
        files = self.run("ls-files --full-name --others --cached --exclude-standard")
        files = files.splitlines()
        return files
```

Compare this with the legacy one https://github.com/conan-io/conan/blob/b70db136a790532a9a80c37068da2c349ef4a0cf/conans/client/tools/scm.py#L232, which didn't include the copy either, and is more fragile.

I think this is way more idiomatic, simple to understand and maintain in both sides, wdyt?

**Comment by Da-LiFe** [comment]
> I think this is way more idiomatic, simple to understand and maintain in both sides, wdyt?

Awesome, I really like the `included_files(self)` idea!

Some of your devs are restricting the export to subfolders, e.g. `src/*` instead of `*` which is easily possible with the `["*", "!.gitignore"]` pattern above. What do you think about having an optional parameter "subdir" which is added to the git cmd with `-C`, like `self.run(f"ls-files -C {subdir} --full-name --others --cached --exclude-standard")`?

This way, one could go for something such as

```python
 def export_sources(self):
       git = Git(self)
       for i in git.included_files("src"):
             shutil.copy2(i, self.export_sources_folder)
       for i in git.included_files("include"):
             shutil.copy2(i, self.export_sources_folder)
```

to export only files in those subdirs, still not exporting the ignored files in those folders.

**Comment by memsharded** [comment]
I think maybe the subfolder case is covered by the already existing constructor argument. This works:
```python
def export_sources(self):
      for i in Git(self, "src").included_files():
          # maybe we want to copy to ``os.path.join(self.export_sources_folder, "src")`` instead
          shutil.copy2(i, self.export_sources_folder)
      for i in Git(self, "include").included_files():
          shutil.copy2(i, self.export_sources_folder)
```

Added a test to the PR https://github.com/conan-io/conan/pull/12246/commits/57bf2c3a145560dd099eb5ffb57161847ccf39b8

**Comment by Da-LiFe** [comment]
> I think maybe the subfolder case is covered by the already existing constructor argument.

Perfekt, I think this is a very nice solution to our use case, thanks! One last question: Is there any argument against the `@property` pattern:

```python
 @property
 def export_sources(self):
       # ask git to provide a list of files to export
       return Git(self).included_files()
```

As you have expressed in the comment above, it might not be obvious were to copy those files with `shutil.copy2`.

**Comment by memsharded** [comment]
> One last question: Is there any argument against the @property pattern:

Yes, we think it makes things a bit confusing, because the idea is:
- Use the attribute ``exports_sources`` for very simple and straigtforward cases
- Use the method ``export_sources()`` for anything not that simple, making things explicit

So when a method is used, the behavior should be more explicit and self-contained. Such a property could seem explicit, but it still relies internally on all the pattern-based copies (yes, the internals of ``exports_sources`` attributes use file walking and pattern matching), making it both more inefficient and complex. Things might work, until there are corner cases like symlinks, hidden files, or other stuff that the pattern matching might handle differently, while the ``shutil.copy2()`` is super simple, explicit and fast.

**Comment by Da-LiFe** [comment]
I made a local test with shutil and I just would like to point out that the example from above is not going to work as it is. So don't put it into the docs :)

```python
 def export_sources(self):
       git = Git(self)
       for i in git.included_files():
             shutil.copy2(i, self.export_sources_folder)
```

Having a repo with the two files `CMakeLists.txt` and `src/source.cpp`, the code above will generate an exported_sources such as `CMakeLists.txt` and `source.cpp` on the same level. Some of the exports_sources attribute magic is still required here to make it work for all cases...

**Comment by memsharded** [comment]
Good catch!

I have extended the test to cover this subfolder. Now the method is a bit more elaborate:
```python
def export_sources(self):
      git = Git(self)
      for i in git.included_files():
          dst =  os.path.join(self.export_sources_folder, i)
          os.makedirs(os.path.dirname(dst), exist_ok=True)
          shutil.copy2(i, dst)
```
I find it a bit annoying that ``shutil.copy`` operations cannot create a subfolder, that seems a limitation of the python API to me, but it is what it is.
Still I think it is still explicit and simple enough, and I find it nice that the ``git`` part and the copy part are decoupled.

**Comment by Da-LiFe** [comment]
> Now the method is a bit more elaborate:

what about using copy instead?

```python
    def export_sources(self):
        from conan.tools.files import copy
        git = Git(self)
        for i in git.included_files():
            copy(self, i, self.recipe_folder, self.export_sources_folder)
```

That's already proposed in the [migration guide](https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html#the-export-sources-method) and seems to be working with subfolders (keep_path).

**Comment by memsharded** [comment]
Yes, both approaches are equally valid IMO. In the above I wanted to remove the "magic" of the ``copy()``, so even if it requires a couple of extra lines, it would be good, and it should work too. And it can be much much faster, which is something that for big projects might make a difference. We can document both approaches, as the ``git.included_files()`` is now decoupled.

**Comment by memsharded** [comment]
Closed by https://github.com/conan-io/conan/pull/12246, will be in beta.4

**Comment by Da-LiFe** [comment]
> Closed by #12246, will be in beta.4

Great, thanks! Is there any chance to get this feature also in `conan.tools.scm.git` for conan 1.x? That would allow us an easier transition.

**Comment by memsharded** [comment]
Oh, yes, sorry, this should have been done directly for 1.X to allow the migration, I will backport it to 1.54
